### PR TITLE
Analysis orchestrator

### PR DESCRIPTION
**Pull request by GirZ0n**
**Сhanges**:
- Added a common `Orchestrator` interface for all orchestrators.
- Now `MultipleAnalysisOrchestrator` uses openers specified in the analyzers.
- Renamed `MultipleAnalysisOrchestrator.kt` to `AnalysisOrchestrator.kt`.
- Add `AnalysisOrchestrator`.
- Wrapped  all executors into `AnalysisOrchestrator`.

### CODE REVIEW

File: lupa-core/src/main/kotlin/org/jetbrains/research/lupa/kotlinAnalysis/AnalysisOrchestrator.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
+package org.jetbrains.research.lupa.kotlinAnalysis
```
Comment (nbirillo) [inline lupa-core/src/main/kotlin/org/jetbrains/research/lupa/kotlinAnalysis/AnalysisOrchestrator.kt:1]: In this Orchestrator version, you have a lot of duplicates of code. Please, change it, e.g. you can consider the following code (I deleted comments, they should be added):

```
/**
 * Classes that inherit from this interface filter the project files by the extension needed to perform the analysis.
 */
abstract class Orchestrator(protected open val executorHelper: ExecutorHelper? = null) {
    abstract fun doAnalysis(projectPath: Path, projectTmpFolderPath: Path): Boolean

    fun execute(projectsDir: Path, outputDir: Path) {
        val tempFolderPath = Paths.get(outputDir.toString(), "tmp")
        tempFolderPath.delete(recursively = true)

        getSubdirectories(projectsDir).forEachIndexed { projectIndex, projectPath ->
            println("Start analyzing project ${projectPath.fileName} (index=$projectIndex)")
            val isSuccessful = doAnalysis(projectPath, tempFolderPath.resolve(projectPath.fileName))
            println("Project ${projectPath.fileName} (index=$projectIndex) " +
                    "was analyzed successfully: $isSuccessful")
            if (isSuccessful) {
                executorHelper?.postExecuteAction(GitRepository(projectPath))
            }
        }

        tempFolderPath.delete()
    }
}

/**
 * Class for execution analyzer for all projects in given directory.
 */
class AnalysisOrchestrator(
    private val analysisExecutor: AnalysisExecutor,
    executorHelper: ExecutorHelper? = null
) : Orchestrator(executorHelper) {

    override fun doAnalysis(projectPath: Path, projectTmpFolderPath: Path): Boolean {
        symbolicCopyOnlyRequiredExtensions(
            fromDirectory = projectPath.toRealPath(),
            toDirectory = projectTmpFolderPath,
            analysisExecutor.requiredFileExtensions,
        )
        val isSuccessful = analysisExecutor.execute(projectTmpFolderPath)
        projectTmpFolderPath.delete(true)
        return isSuccessful
    }
}

/**
 * Class for execution multiple analyzers for all projects in given directory.
 */
class MultipleAnalysisOrchestrator(
    analysisExecutors: List<AnalysisExecutor>,
    executorHelper: ExecutorHelper? = null
) : Orchestrator(executorHelper) {

    private val analyzersByExtensions = analysisExecutors.groupBy { it.requiredFileExtensions }

    override fun doAnalysis(projectPath: Path, projectTmpFolderPath: Path): Boolean {
        return analyzersByExtensions.map { (extensions, analyzers) ->
            symbolicCopyOnlyRequiredExtensions(
                fromDirectory = projectPath.toRealPath(),
                toDirectory = projectTmpFolderPath,
                extensions
            )

            val analyzersByOpener = analyzers.groupBy { it.repositoryOpener }
            val isSuccessful = analyzersByOpener.map { (opener, analyzers) ->
                MultipleAnalysisExecutor(
                    analyzers,
                    repositoryOpener = opener
                ).execute(projectTmpFolderPath)
            }.all { it }

            projectTmpFolderPath.delete(true)
            isSuccessful
        }.all { it }
    }
}
```

Lets discuss it in DM

**Review by nbirillo — CHANGES_REQUESTED** [review]
Please, avoid code duplications

**Comment by GirZ0n** [comment]
@nbirillo. can you please merge this PR?

**Review by nbirillo — APPROVED** [review]
Great! Thank you a lot!

**Pull request merged**